yarn-project/end-to-end/src/e2e_escrow_contract.test.ts

```diff
@@ -0,0 +1,115 @@
+    const transferAbi = ZkTokenContractAbi.functions.find(f => f.name === 'transfer')!;
+    const transferSelector = generateFunctionSelector(transferAbi.name, transferAbi.parameters);
+    expect(transferSelector).toEqual(toBufferBE(0xdcd4c318n, 4));
+  });
+
+  beforeEach(async () => {
+    // Setup environment
+    ({ aztecNode, aztecRpcServer, accounts, wallet, logger } = await setup(2));
+    [owner, recipient] = accounts;
+
+    // Generate private key for escrow contract, register key in rpc server, and deploy
+    // Note that we need to register it first if we want to emit an encrypted note for it in the constructor
+    // TODO: We need a nicer interface for deploying contracts!
+    escrowPrivateKey = PrivateKey.random();
+    escrowPublicKey = await generatePublicKey(escrowPrivateKey);
+    const salt = Fr.random();
+    const deployInfo = await getContractDeploymentInfo(EscrowContractAbi, [owner], salt, escrowPublicKey);
+    await aztecRpcServer.addAccount(escrowPrivateKey, deployInfo.address, deployInfo.partialAddress);
+    const escrowDeployTx = EscrowContract.deployWithPublicKey(aztecRpcServer, escrowPublicKey, owner);
+    await escrowDeployTx.send({ contractAddressSalt: salt }).wait();
+    escrowContract = new EscrowContract(escrowDeployTx.completeContractAddress!, wallet);
+    logger(`Escrow contract deployed at ${escrowContract.address}`);
+
+    // Deploy ZK token contract and mint funds for the escrow contract
+    zkTokenContract = await ZkTokenContract.deploy(aztecRpcServer, 100n, escrowContract.address)
+      .send()
+      .wait()
+      .then(r => new ZkTokenContract(r.contractAddress!, wallet));
+    logger(`Token contract deployed at ${zkTokenContract.address}`);
+  }, 100_000);
+
+  afterEach(async () => {
+    await aztecNode?.stop();
+    if (aztecRpcServer instanceof AztecRPCServer) await aztecRpcServer.stop();
+  });
+
+  const expectBalance = async (who: AztecAddress, expectedBalance: bigint) => {
+    const [balance] = await zkTokenContract.methods.getBalance(who).view({ from: who });
+    logger(`Account ${who} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  it('withdraws funds from the escrow contract', async () => {
+    await expectBalance(owner, 0n);
+    await expectBalance(recipient, 0n);
+    await expectBalance(escrowContract.address, 100n);
+
+    logger(`Withdrawing funds from token contract to ${recipient}`);
+    await escrowContract.methods.withdraw(zkTokenContract.address, 30, recipient).send().wait();
+
+    await expectBalance(owner, 0n);
+    await expectBalance(recipient, 30n);
+    await expectBalance(escrowContract.address, 70n);
+  }, 60_000);
+
+  it('refuses to withdraw funds as a non-owner', async () => {
+    await expect(
+      escrowContract.methods.withdraw(zkTokenContract.address, 30, recipient).simulate({ origin: recipient }),
+    ).rejects.toThrowError();
+  }, 60_000);
```

Comment:
I'd've expected this test to fail with a `cannot satisfy constraints`, but I got this error instead. For context, I'm running an `ensure_note_exists` for a note which does not exist. Is it possible to catch this somehow? cc @LeilaWang 

```
  wasm Error(2018): private data tree root mismatch at read_request[0]
  wasm 	expected root:    0x01834dbe674ccaa98f5aaafdcd8a42267503a0c141f582ad7f53564fab450c23
  wasm 	but got root*:    0x23b12bc80a094747ce77a30b71288abcb7cbf648a8891cfe907405db219c8549
  wasm 	read_request:     0x156409eef4798a1db5b3a8d73dfa047c3c0de0f06fc888209b1a191c31b5dde8
  wasm 	siloed-rr (leaf): 0x1115a070391204040030b3f5e4ab5d5e6fc5b3c82e16be7de43f2b089961e7c6
  wasm 	leaf_index: 0x0000000000000000000000000000000000000000000000000000000000000040
  wasm 	is_transient: 0
  wasm 	hint_to_commitment: 0x0000000000000000000000000000000000000000000000000000000000000000
  wasm 	* got root by siloing read_request (compressing with storage_contract_address to get leaf) and merkle-hashing to a root using membership witness (mem: 34.50MB) +30s
```